internal/metrics/metrics.go

```diff
@@ -0,0 +1,114 @@
+package metrics
+
+
+import (
+	"time"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	ms = 0.001
+	s  = 1000 * ms
+	m  = 60 * s
+	kb = 1024
+	mb = 1024*kb
+)
+
+type metrics struct {
+	registry *prometheus.Registry
+	endpoint string
+
+	// Custom metics
+	request_durations *prometheus.HistogramVec
+	response_sizes    *prometheus.HistogramVec
+
+}
+
+func (m *metrics) GetEndpoint() string {
+	return m.endpoint
+}
+
+/** Create a new metric instance
+ *
+ * The metrics are applies to a new non-global prometheus registry
+ */
+func NewMetrics(endpoint string) *metrics {
+	registry := prometheus.NewRegistry()
+	metrics := &metrics{
+		registry: registry,
+		endpoint: strings.TrimLeft(endpoint, "/"),
+
+		request_durations: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "vdsslice_durations_histogram_seconds",
+			Help:    "VDSslice latency distributions.",
+			Buckets: []float64{100*ms, 500*ms, 1*s, 2*s, 5*s, 20*s, 1*m, 2*m},
+		}, []string{"method", "path", "status", "cachehit"}),
+
+		response_sizes: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "vdsslice_response_sizes_histogram_bytes",
+			Help:    "VDSslice response size distributions.",
+			Buckets: []float64{100*kb, 1*mb, 5*mb, 10*mb, 20*mb, 50*mb, 100*mb, 200*mb},
+		}, []string{"method", "path", "status"}),
+	}
+
+	registry.MustRegister(metrics.request_durations)
+	registry.MustRegister(metrics.response_sizes)
+
+	return metrics;
+}
```

Comment:
Do you know what happens with old data?
Saw somewhere it is removed after 15 days, but it is for Prometheus servers.
Have you maybe seen something like that for our case? Are metrics as long lived as Radix replica?

Reply:
Yes I believe so. But logs are cleared on crashes and restarts. But it doesn't matter. A prometheus metric is only a snapshot in time. I.e. "How many request has been processed now". It's graphana that makes the history. It reads the "snapshots" every X seconds and use the difference to provide trends over time. So hopefully it shouldn't matter if logs are reset.